src/Optimizer/Transformer/MinifyHtml.php

```diff
@@ -0,0 +1,248 @@
+        } elseif ($node instanceof DOMComment) {
+            $nodesToRemove = $this->minifyCommentNode($node);
+        } elseif ($node instanceof Element && $node->tagName === Tag::SCRIPT) {
+            $this->minifyScriptNode($node, $errors);
+        }
+
+        // Update options based on the current node.
+        if (isset($node->tagName)) {
+            if ($canCollapseWhitespace && !$this->canCollapseWhitespace($node->tagName)) {
+                $canCollapseWhitespace = false;
+            }
+
+            if ($node->tagName === Tag::HEAD || $node->tagName === Tag::HTML) {
+                $inBody = false;
+            } elseif ($node->tagName === Tag::BODY) {
+                $inBody = true;
+            }
+        }
+
+        if ($node->hasChildNodes()) {
+            foreach ($node->childNodes as $childNode) {
+                $nodesToRemove = array_merge(
+                    $nodesToRemove,
+                    $this->minifyNode($childNode, $errors, $canCollapseWhitespace, $inBody)
+                );
+            }
+        }
+
+        return $nodesToRemove;
+    }
+
+    /**
+     * Minify a text type DOM node.
+     *
+     * @param DOMText $node                  Text to apply the transformations to.
+     * @param bool    $canCollapseWhitespace Optional. Whether whitespace can be collapsed. Defaults to true.
+     * @param bool    $inBody                Optional. Whether the node is in the body. Defaults to false.
+     * @return DOMNode[] Array of nodes to be removed.
+     */
+    private function minifyTextNode(DOMText $node, $canCollapseWhitespace = true, $inBody = false)
+    {
+        if (! $node->data || ! $this->configuration->get(MinifyHtmlConfiguration::COLLAPSE_WHITESPACE)) {
+            return [];
+        }
+
+        if ($canCollapseWhitespace) {
+            $node->data = $this->normalizeWhitespace($node->data);
+        }
+
+        if (! $inBody) {
+            $node->data = trim($node->data);
+        }
+
+        // Remove empty nodes.
+        if (strlen($node->data) === 0) {
+            return [$node];
+        }
+
+        return [];
+    }
+
+    /**
+     * Minify/remove a comment node.
+     *
+     * @param DOMComment $node Comment to apply the transformations to.
+     * @return DOMNode[] Array of nodes to be removed.
+     */
+    private function minifyCommentNode(DOMComment $node)
+    {
+        if (! $node->data || ! $this->configuration->get(MinifyHtmlConfiguration::REMOVE_COMMENTS)) {
+            return [];
+        }
+
+        $commentIgnorePattern = $this->configuration->get(MinifyHtmlConfiguration::COMMENT_IGNORE_PATTERN);
+        if (! empty($commentIgnorePattern) && preg_match($commentIgnorePattern, $node->data)) {
+            return [];
+        }
+
+        // In case the main $document has `securedDoctype`.
+        if (preg_match('/^amp-doctype html/i', $node->data)) {
+            return [];
+        }
```

Comment:
There's another case that I think may have been missed here, and that is the Mustache template workaround for HTML tables which is documented at https://amp.dev/documentation/components/amp-mustache/#tables

![image](https://user-images.githubusercontent.com/134745/124337997-68feca00-db5a-11eb-97da-12d18a574cf6.png)

So if a comment node appears inside of a `template[type=amp-mustache]` element, and the comment text includes `{{`, then it should be ignored.

Reply:
@ediamin Maybe you could open a new issue or PR to address this point?

Reply:
Er, I just filed an issue for it: #285.

Reply:
@westonruter Yeah sure.

Reply:
I see you've already created the issue. Thank you very much @westonruter.